### PR TITLE
support distributed query profiling

### PR DESCRIPTION
**Pull request by zhouruiapple**
…thread #173

assign aggregation profiling to cloned profiling.

### CODE REVIEW

**Comment by zhouruiapple** [comment]
There is still a issue of race condiation, which will be solved later.

File: qpmodel/ExprSubquery.cs [hunk 1]
```diff
@@ -261,25 +261,28 @@ public override void Bind(BindContext context)
 
         public override Value ExecNonDistributed(ExecContext context, Row input)
         {
-            Debug.Assert(type_ != null);
-            Value expr = expr_().Exec(context, input);
+            lock(this)
```
Comment (zhouqingqing) [inline qpmodel/ExprSubquery.cs:264]: why this change? Do we already handled this in Exec()?

File: qpmodel/stmt.cs [hunk 2]
```diff
@@ -131,6 +131,13 @@ public virtual List<Row> Exec()
             {
                 Console.WriteLine(physicPlan_.Explain(queryOpt_.explain_));
             }
+
+            var dcontext = context as DistributedContext;
+            if (dcontext != null)
+            {
+                dcontext.machines_.WaitThreads();
```
Comment (zhouqingqing) [inline qpmodel/stmt.cs:138]: why we need to waitThreads() here?

File: qpmodel/ExprSubquery.cs (hunk 1 above)
Comment (zhouruiapple) [reply qpmodel/ExprSubquery.cs:264]: The mainline code also fixes the bug. Keep the bug fix code in this commit the when merging code.
I will remove above code.

File: test/regress/expect/tpch0001_d/q01.txt [hunk 3]
```diff
@@ -29,7 +29,7 @@ PhysicOrder  (inccost=71071.35, cost=11.35, rows=6, memory=372) (actual rows=4)
         Group by: l_returnflag[0], l_linestatus[1]
         -> PhysicGather 1 : 10 (inccost=65135, cost=59130, rows=5913) (actual rows=5914)
             Output: l_returnflag[8],l_linestatus[9],l_quantity[4],l_extendedprice[5],l_extendedprice[5]*1-l_discount[6],1-l_discount[6],1,l_discount[6],l_extendedprice[5]*1-l_discount[6]*1+l_tax[7],1+l_tax[7],l_tax[7]
-            -> PhysicScanTable lineitem (inccost=6005, cost=6005, rows=5913) (actual rows=0)
+            -> PhysicScanTable lineitem (inccost=6005, cost=6005, rows=5913) (actual rows=591, loops=10)
```
[the remark this answers is not in the document]
Comment (zhouruiapple) [reply test/regress/expect/tpch0001_d/q01.txt:32]: Got it. Can you create a issue for this change?

File: qpmodel/stmt.cs (hunk 2 above)
Comment (zhouruiapple) [reply qpmodel/stmt.cs:138]: If the query plan(in LISP) as (PhysicGather(...(PhysicGather))), we must wait all threads to guarantee loop count of second PhysicGather is fixed value.

File: qpmodel/PhysicalNode.cs [hunk 4]
```diff
@@ -1697,17 +1707,33 @@ public override void Exec(Action<Row> callback)
...
+                    {
+                        aggProfiling_.nrows_++;
+                    }
                     callback(l);
                 }
             });
         }
 
+        // assign aggregation profiling to cloned profiling.
+        public override PhysicNode Clone()
```
[the remark this answers is not in the document]
Comment (zhouruiapple) [reply qpmodel/PhysicalNode.cs:1746]: Query plan under Gather with actual "actual rows = 0" because of all Profilings under Gather are cloned. So we should keep reference of original Profiling in all cloned profilings.

All expected results are updated.

File: qpmodel/Statis.cs [hunk 5]
```diff
@@ -346,7 +346,8 @@ public void ComputeStats(int index, List<Row> samples)
                 foreach (var g in groups)
                     sortgroup.Add(g.Key, g.Count());
 
-                var sorted = from pair in sortgroup orderby pair.Value descending select pair;
+                // use top 100 values to calculate frequency, ensure that pairs are sorted in a fixed order.
+                var sorted = from pair in sortgroup orderby pair.Value descending, pair.Key descending select pair;
```
Comment (zhouruiapple) [inline qpmodel/Statis.cs:350]: we must keep the order of @sorted. Sort by @pair.Value, and then sort by @pair.Key. Otherwise top 100 of @sorted cant not be guaranteed.
